### DISCUSSION TITLE
Infer `SendType` in a Python Generator (Coroutine)

### DISCUSSION THREAD

**Discussion by Tomperez98** [microsoft/pyright#9576] [Q&A] [answered]
I have a very simple use case. When yielding a either `LFI[T] | RFI[T]` I'll get send back a `Promise[T]` and when yielding a `Promise[T]` I'll get send back a value `T`. 

It seems that it's not possible for the type system to infer the send type based on the yielded type. But I don't know if there's anything I'm missing and I could be using.

```py
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Generator

    from resonate.context import Context
    from resonate.record import Promise
    from resonate.typing import Yieldable


def bar(ctx: Context, n: int) -> str: ...
def foo(ctx: Context, n: int) -> Generator[Yieldable, Any, str]:
    p: Promise[str] = yield ctx.lfi(bar, n)
    v: str = yield p
    return v
```

I currently rely on annotating variable type on assignment. At annotate the Coroutine as `Generator[Yieldable, Any, T]`. Would it be possible to do otherwise? Not needing to annotate variables inside the coroutine

**Comment by erictraut** [comment]
Could you provide a self-contained minimal code sample or tell me what library I need to install to repro your code? Your sample references the symbol `resonate` which isn't imported. I tried `pip install resonate`, but I don't think that's the right library.

The send type for a generator function is specified as the second type argument when specializing the `Generator` return type. You've specified the send type as `Any` in your code sample. If your intent is for it to be something more specific like `str`, you can specify it as `Generator[Yieldable, str, str]`.

**Reply by Tomperez98** [reply]
Sorry. Yes, resonate can be installed with `pip install resonate-sdk`.

I get that I could specify the second type (send) not to be `Any`. But I need is more like dependent type.

Basically if I yield and `str` I'll receive an `int` and if I yield an `int` I'll receive a `str`. I don't know if that's possible

**Reply by erictraut** [reply, marked as answer]
Pyright uses the specified SendType from the return type. If you are using multiple different send types in your generator (which is pretty atyipcal), you'd need to declare the type on the receiving variable like you've done in your example.

**Reply by Tomperez98** [reply]
Thank you so much!
